### ISSUE TITLE
ES v6.7.1 Terms lookup mechanism filter inside dis_max query return error: unsupported_operation_exception: query must be rewritten first

### ISSUE THREAD

**Issue by hrsvrma** [elastic/elasticsearch#40953]
**Elasticsearch version** (`bin/elasticsearch --version`):
v6.7.1
v6.5.4

**Plugins installed**: []

**JVM version** (`java -version`):
hosted on cloud.elastic.co

**OS version** (`uname -a` if on a Unix-like system):
hosted on aws cloud.elastic.co

**Description of the problem including expected versus actual behavior**:

Dismax query with terms lookup mechanism filter returns 400 Bad request: unsupported_operation_exception: query must be rewritten first (full response below)

**Steps to reproduce**:

Elasticsearch v6.7.1, v6.5.4 hosted on cloud.elastic.co

Steps to reproduce

**1. create index *users***
```
PUT /users
{
  "mappings": {
    "_doc" : {
      "properties" : {
        "firstName" : {
          "type" : "text",
          "analyzer":"standard"
        },
        "lastName" : {
          "type" : "text",
          "analyzer":"standard"
        },
        "following" : {
          "type": "keyword"
        }
      }
    }
  }
}
```

**2. add *users* docs**
```
PUT /users/_doc/user1
{
  "firstName": "foo1",
  "lastName": "bar1",
  "following": ["user1", "user2", "1", "dave"]
}
```

**3. create index *feeds***
```
PUT /feeds
{
  "mappings": {
    "_doc" : {
      "properties" : {
        "userId" : {
          "type": "keyword"
        },
        "feedType" : {
          "type": "keyword"
        }
      }
    }
  }
}
```

**4. add feeds docs**
```
PUT /feeds/_doc/1
{
  "userId": "1",
  "feedType": "feed1"
}
```

```
PUT /feeds/_doc/feed2
{
  "userId": "user2",
  "feedType": "feed2"
}
```

**5. run simple filter query with *terms lookup mechanism*** (This works fine)
```
POST /feeds/_search
{
  "query": {
    "bool": {
      "must": {
        "match_all": {}
      },
      "filter": {
        "bool": {
          "must": [
            {
              "terms": {
                "userId": {
                  "index": "users",
                  "type": "_doc",
                  "id": "user1",
                  "path": "following"
                }
              }
            }
          ]
        }
      }
    }
  }
}
```
Success Response
```
200 OK
{
  "hits": {
    "hits": [
      {
        "_score": 1,
        "_type": "_doc",
        "_id": "feed2",
        "_source": {
          "feedType": "feed2",
          "userId": "user2"
        },
        "_index": "feeds"
      },
      {
        "_score": 1,
        "_type": "_doc",
        "_id": "1",
        "_source": {
          "feedType": "feed1",
          "userId": "1"
        },
        "_index": "feeds"
      }
    ],
    "total": 2,
    "max_score": 1
  },
  "_shards": {
    "successful": 5,
    "failed": 0,
    "skipped": 0,
    "total": 5
  },
  "took": 6,
  "timed_out": false
}
```
**6. run *dis_max query* with terms lookup mechanism**
```
POST /feeds/_search
{
  "query": {
    "dis_max": {
      "tie_breaker": 0.7,
      "boost": 1.2,
      "queries": [
        {
          "match_all": {}
        },
        {
          "terms": {
                "userId": {
                  "index": "users",
                  "type": "_doc",
                  "id": "user1",
                  "path": "following"
                }
              }
        }
      ]
    }
  }
}
```
**Error Response**
```
400 Bad Request
{
  "status": 400,
  "error": {
    "failed_shards": [
      {
        "node": "bOTWWzDfQFO6tzdl4aHQUQ",
        "index": "feeds",
        "reason": {
          "index_uuid": "q2DlqKmcRfqSUUWSJUU4-A",
          "index": "feeds",
          "reason": "failed to create query: {\n  \"dis_max\" : {\n    \"tie_breaker\" : 0.7,\n    \"queries\" : [\n      {\n        \"match_all\" : {\n          \"boost\" : 1.0\n        }\n      },\n      {\n        \"terms\" : {\n          \"userId\" : {\n            \"index\" : \"users\",\n            \"type\" : \"_doc\",\n            \"id\" : \"user1\",\n            \"path\" : \"following\"\n          },\n          \"boost\" : 1.0\n        }\n      }\n    ],\n    \"boost\" : 1.2\n  }\n}",
          "caused_by": {
            "reason": "query must be rewritten first",
            "type": "unsupported_operation_exception"
          },
          "type": "query_shard_exception"
        },
        "shard": 0
      }
    ],
    "root_cause": [
      {
        "index_uuid": "q2DlqKmcRfqSUUWSJUU4-A",
        "index": "feeds",
        "reason": "failed to create query: {\n  \"dis_max\" : {\n    \"tie_breaker\" : 0.7,\n    \"queries\" : [\n      {\n        \"match_all\" : {\n          \"boost\" : 1.0\n        }\n      },\n      {\n        \"terms\" : {\n          \"userId\" : {\n            \"index\" : \"users\",\n            \"type\" : \"_doc\",\n            \"id\" : \"user1\",\n            \"path\" : \"following\"\n          },\n          \"boost\" : 1.0\n        }\n      }\n    ],\n    \"boost\" : 1.2\n  }\n}",
        "type": "query_shard_exception"
      }
    ],
    "grouped": true,
    "reason": "all shards failed",
    "phase": "query",
    "type": "search_phase_execution_exception"
  }
}
```

**logs (if relevant)**:
[es.log](https://github.com/elastic/elasticsearch/files/3054152/es.log)

**Comment by hrsvrma** [comment]
Hi @jimczi ,

Thanks for the fix. Now that the pull request is merged, when can this be expected to be available publicly in build?

**Comment by hrsvrma** [comment]
Hi @jimczi,

Thanks for the fix. Is this fix available in released v7.0.0?

**Comment by jimczi** [comment]
> Thanks for the fix. Is this fix available in released v7.0.0?

No but it will be available in the next release which could be 7.0.1 if we need a patch release for 7.0 or 7.1.0 otherwise. It is also pushed to 6.7 so it will also be available in the next patch release of this version.

**Comment by nirmalc** [comment]
Any chance that this defect will be back ported to 5.x ? @jimczi .
